src/cockpit/protocol.py

```diff
@@ -84,20 +86,21 @@ def frame_received(self, frame):
             logger.debug('data received: %d bytes of data for channel %s', len(data), channel)
             self.channel_data_received(channel, data)
         else:
-            message = json.loads(data)
             try:
-                command = message['command']
-            except KeyError as exc:
+                message = JsonObject(json.loads(data))
+                command = message.get_str('command')
+
+                channel = message.get_str('channel', None)
+                if channel is not None:
+                    logger.debug('channel control received %s', message)
+                    self.channel_control_received(channel, command, message)
+                else:
+                    logger.debug('transport control received %s', message)
+                    self.transport_control_received(command, message)
+
+            except JsonError as exc:
                 raise CockpitProtocolError('control message is missing command field') from exc
```

Comment:
Is that the generic replacement for the specific rules like the former
```py
        if max_read_size is not None and not isinstance(max_read_size, int):
            raise ChannelError('protocol-error', message='max_read_size must be an integer')
```

This seems rather vague, then. It can't tell apart the "missing" from the "wrong type" situations, and does not even give the field name. Please handle Json{Key,Type} with different error messages. Also, could these exceptions grow `.key` (or `.name`) and `.expected_type` fields, so that the error message can include them?